addons/repair/tests/test_repair.py

```diff
@@ -618,3 +619,54 @@ def test_repair_with_product_in_package(self):
         repair_order.move_ids.quantity = 1
         repair_order.action_repair_end()
         self.assertEqual(repair_order.state, 'done')
+
+    def test_repair_multi_unit_order_with_serial_tracking(self):
+        """
+        create a repair order tied to a sale order with multiple units which have serial number tracking enabled
+        """
+        company_a = self.env['res.company'].create({'name': 'CompanyA'})
+        company_b = self.env['res.company'].create({'name': 'CompanyB'})
+
+        product = self.env['product.product'].create({
+            'name': 'productA',
+            'detailed_type': 'product',
+            'tracking': 'serial',
+            'create_repair': True,
+        })
+
+        stock_location = self.env['stock.location'].create({
+            'name': 'stock_location',
+            'usage': 'internal',
+            'company_id': company_a.id,
+        })
+
+        lots = []
+        for i in range(3):
+            lots.append(self.env['stock.lot'].create({
+                'name': str(i),
+                'product_id': product.id,
+                'company_id': company_a.id,
+            }))
+            self.env['stock.quant'].create({
+                'product_id': product.id,
+                'location_id': stock_location.id,
+                'inventory_quantity': 1.0,
+                'lot_id': lots[i].id,
+            })._apply_inventory()
+
+        sale_order = self.env['sale.order'].create({
+            'company_id': company_a.id,
+            'partner_id': company_b.id,
+            'order_line': [Command.create({
+                'product_id': product.id,
+                'product_uom_qty': 3.0,
+            }),]
+        })
+        sale_order.action_confirm()
+
+        repair_order = self.env['repair.order'].search([('sale_order_id', '=', sale_order.id)])
+        repair_order.write({
+            'lot_id': lots[0].id,
+        })
+        repair_order.action_repair_start()
+        repair_order.action_repair_end()
```

Comment:
Few comments:
\- It's always a good practice to use existing data in `self`
\- Make things simple: here, companies/location creation is useless. And, except if I'm missing something, the only issue is the RO creation: we should have 3 RO instead of one. No need to create any lot/quant or to process the ROs
\- Create in batch as much as possible:
```py
for i in range(3):
    self.env['stock.lot'].create({...})
# should be
self.env['stock.lot'].create([{...} for i in range(3)])
```
\- `'partner_id': company_b.id` 
Watch out for field definition. Here `partner_id` needs a `res.partner`, not a `res.company`:
https://github.com/odoo/odoo/blob/2c5a8568507bcabbf244abddb8e8ca29b1494a4f/addons/sale/models/sale_order.py#L61-L62
\- If you really need to use a `search`, adding a `limit` is a good practice too (perf matter). But actually, here we have a field on the SO that allows us to directly get the ROs. (Implicitly, it also checks that all records are correctly linked)

So, it could be something like this
```py
        product = self.env['product.product'].create({
            'name': 'productA',
            'detailed_type': 'product',
            'tracking': 'serial',
            'create_repair': True,
        })

        sale_order = self.env['sale.order'].create({
            'partner_id': self.res_partner_1.id,
            'order_line': [Command.create({
                'product_id': product.id,
                'product_uom_qty': 3.0,
            })]
        })
        sale_order.action_confirm()

        repair_orders = sale_order.repair_order_ids
        self.assertRecordValues(repair_orders, [
            {'product_id': product.id, 'product_qty': 1.0},
            {'product_id': product.id, 'product_qty': 1.0},
            {'product_id': product.id, 'product_qty': 1.0},
        ])
```